lib/tree.rb

```diff
@@ -16,40 +16,110 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+# add helper 
+  def helper_add(current, new_node)
+    return new_node if current.nil? 
+
+    if new_node.key <= current.key 
+      current.left = helper_add(current.left , new_node)
+    else   
+      current.right = helper_add(current.right, new_node)
+    end 
+    return current
+  end 
+
+  # Time Complexity: log(n)
+  # Space Complexity: O(1)
   def add(key, value)
```

Comment:
👍 Since this is recursion, you have the system stack and the space complexity is O(log n) if the tree is balanced.